Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+                // other stream is invalid and MUST be ignored.
+                return nil
+            }
+            
+            var origins: [String] = []
+            var remaining = header.length
+            while remaining > 0 {
+                guard remaining >= 2 else {
+                    // If less than two bytes remain, this is a malformed frame.
+                    throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+                }
+                let originLen: UInt16 = bytes.readInteger()!
+                remaining -= 2
+                
+                guard remaining >= Int(originLen) else {
+                    // Malformed frame.
+                    throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+                }
+                let origin = bytes.readString(length: Int(originLen))!
+                remaining -= Int(originLen)
+                
+                origins.append(origin)
+            }
+            
+            payload = .origin(origins)
+            
+        default:
+            // RFC 7540 § 4.1 https://httpwg.org/specs/rfc7540.html#FrameHeader
+            //    "Implementations MUST ignore and discard any frame that has a type that is unknown."
+            self.state = .idle
+            bytes.moveReaderIndex(forwardBy: header.length)
+            return nil
+        }
+        
+        self.state = .idle
+        let frame = HTTP2Frame(streamID: streamID, flags: flags, payload: payload)
+        
+        switch payload {
+        case .headers where !flags.contains(.endHeaders), .pushPromise where !flags.contains(.endHeaders):
+            // don't emit the frame, store it and wait for CONTINUATION frames to arrive
+            self.currentHeaderFrame = frame
+            return nil
+        default:
+            return frame
+        }
+    }
+    
+    private func parseContinuation(length: Int, streamID: HTTP2StreamID, flags: HTTP2Frame.FrameFlags, bytes: inout ByteBuffer) throws -> HTTP2Frame? {
+        // CONTINUATION frame : RFC 7540 § 6.10
+        guard var frame = self.currentHeaderFrame else {
+            // A CONTINUATION frame MUST be preceded by a HEADERS, PUSH_PROMISE or
+            // CONTINUATION frame without the END_HEADERS flag set. A recipient that
+            // observes violation of this rule MUST respond with a connection error
+            // (Section 5.4.1) of type PROTOCOL_ERROR.
+            throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+        }
+        
+        let stream = HTTP2StreamData.getStream(for: streamID, allocator: self.allocator)
+        
+        // this just contains a header block fragment and flags
+        let headers: HPACKHeaders
+        if bytes.readableBytes == length {
+            headers = try stream.decoder.decodeHeaders(from: &bytes)
+        } else {
+            var slice = bytes.readSlice(length: length)!
+            headers = try stream.decoder.decodeHeaders(from: &slice)
```

Comment:
This logic is not quite right. A header block is decompressed as a single unit, even if it's split across multiple CONTINUATION frames. Essentially, you can think of CONTINUATION frames as a special jumbo frame abstraction solely for HEADERS.

The TL;DR here is that you have to buffer the frame parts all together and then decompress them as one.